### PR TITLE
 Enable Split when the partitioning dimension is unknown

### PR DESCRIPTION
**Pull request by yuanbyu**

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -68,15 +68,16 @@ OPERATOR_SCHEMA(Split)
...
             "Constrain input types to float tensors.")
-    .Attr("axis",
-          "Which axis to split on",
-          AttrType::INT)
-    .Attr("split",
-          "length of each output",
-          AttrType::INTS)
+    .Attr("axis", "Which axis to split on", AttrType::INT)
+    .Attr("split", "Optional, lengths for the outputs", AttrType::INTS)
+    .Attr("num_split", "Optional, the number of outputs.", AttrType::INTS)
```
Comment (ebarsoum) [inline onnx/defs/tensor/defs.cc]: Shouldn't the type be AttrType::INT?

File: docs/Operators.md [hunk 2]
```diff
@@ -4511,8 +4511,12 @@ opset_import {
 ### <a name="Split"></a><a name="split">**Split**</a>
 
   Split a tensor into a list of tensors, along the specified
-  'axis'. Lengths of the parts can be specified using argument 'split'.
-  Otherwise, the tensor is split to equal sized parts.
+  'axis'. Either 'split' or 'num_split' must be defined. If 'split'
```
Comment (ebarsoum) [inline docs/Operators.md]: This is breaking change, you should bump the version number.

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (yuanbyu) [reply onnx/defs/tensor/defs.cc]: Thanks.

File: docs/Operators.md (hunk 2 above)
Comment (yuanbyu) [reply docs/Operators.md]: Yes. Done.

**Comment by dzhulgakov** [comment]
Hm, the original op description was just relying on the number of outputs specified? Does it mean that we don't want to rely on the # outputs for ops behavior and instead make number of outputs fully inferable from the attributes?

I'm fine with it, just double-checking

**Comment by bddppq** [comment]
This looks to me is really an unnecessary breaking change. The number of outputs is already there, just count the outputs. Adding another attributes for specifying it seems redundant. More importantly, we started with this, published v1, we really shouldn't frequently make breaking changes for no good reason.

**Comment by yuanbyu** [comment]
@dzhulgakov @bddppq Thinking about it again, I think you are right. I got confused with how to do graph construction from CoreML.  I will cancel it.

**Comment by yuanbyu** [comment]
Thinking about it again, I think that this PR actually addresses some use cases that the current Split can't. Suppose we want to split x evenly into 4 parts on axis=0 but x.dim(0) is unknown, without num_splits I don't see how we can accomplish it. I am reopening the PR.

**Comment by yuanbyu** [comment]
@dzhulgakov @bddppq PTAL.

**Comment by bddppq** [comment]
@yuanbyu The idea is to treat the number of outputs in this Operator as num_splits.

**Comment by yuanbyu** [comment]
@bddppq Sorry I was confused again. This is not needed. Thanks to @linkerzhang for explaining it.

**Pull request closed without merging**